File: test_translations_e2e.py
#!/usr/bin/env python3
"""E2E test for Spanish translations in FamilyFinance"""

import http.server
import threading
import time
import os
import sys

# Start server
os.chdir(os.path.dirname(os.path.abspath(__file__)) + "/static")
server = http.server.HTTPServer(
    ("localhost", 8765), http.server.SimpleHTTPRequestHandler
)
t = threading.Thread(target=server.serve_forever)
t.daemon = True
t.start()
time.sleep(1)

# Read the HTML file
html_file = os.path.dirname(os.path.abspath(__file__)) + "/static/index.html"
with open(html_file, "r") as f:
    html = f.read()

print("=" * 60)
print("FamilyFinance E2E Translation Tests")
print("=" * 60)

tests_passed = 0
tests_failed = 0


def test(name, condition):
    global tests_passed, tests_failed
    if condition:
        print(f"  ✅ {name}")
        tests_passed += 1
    else:
        print(f"  ❌ {name}")
        tests_failed += 1


# Test 1: Verify translations.es block exists
print("\n1. Translation objects exist:")
test("translations.es block exists", "es: {" in html)
test("translations.en block exists", "en: {" in html)
test("bilingualTranslations exists", "const bilingualTranslations" in html)

# Test 2: Verify key Spanish translations exist in es block
print("\n2. Key Spanish translations in es:")
es_keys = {
    "new_card": "Nueva Tarjeta",
    "new_service": "Nuevo Servicio",
    "add_income": "Agregar Ingreso",
    "expenses": "Gastos",
    "income": "Ingresos",
    "debt": "Deuda",
    "save": "Guardar",
    "loading": "Cargando",
    "password": "Contraseña",
    "enter": "Entrar",
    "summary": "Resumen",
    "reports": "Reportes",
    "budgets": "Presupuestos",
    "ai_assistant": "Asistente IA",
    "kakebo": "Kakebo",
    "pay_immediately": "Pagar de Inmediato",
    "pay_overdue": "Pagar Vencido",
    "pay_early": "Pagar Adelantado",
}

for key, expected in es_keys.items():
    # Check in translations.es block (between 'es: {' and 'en: {')
    es_start = html.find("es: {")
    en_start = html.find("en: {")
    es_block = html[es_start:en_start]
    test(f"  {key} = '{expected}'", expected in es_block)

# Test 3: Verify English translations exist in en block
print("\n3. Key English translations in en:")
en_keys = {
    "new_card": "New Card",
    "new_service": "New Service",
    "add_income": "Add Income",
    "expenses": "Expenses",
    "income": "Income",
    "debt": "Debt",
    "save": "Save",
    "loading": "Loading",
    "password": "Password",
    "enter": "Enter",
    "summary": "Summary",
    "reports": "Reports",
    "budgets": "Budgets",
    "ai_assistant": "AI Assistant",
    "kakebo": "Kakebo",
    "pay_immediately": "Pay Immediately",
    "pay_overdue": "Pay Overdue",
    "pay_early": "Pay Early",
}

en_start = html.find("en: {")
en_block = html[en_start : en_start + 10000]  # Get a chunk of en block

for key, expected in en_keys.items():
    test(f"  {key} = '{expected}'", expected in en_block)

# Test 4: Verify data-i18n attributes exist in HTML
print("\n4. data-i18n attributes in HTML:")
data_i18n_checks = [
    'data-i18n="new_card"',
    'data-i18n="new_service"',
    'data-i18n="add_income"',
    'data-i18n="expenses"',
    'data-i18n="income"',
    'data-i18n="debt"',
    'data-i18n="save"',
    'data-i18n="loading"',
    'data-i18n="password"',
    'data-i18n="enter"',
    'data-i18n="summary"',
    'data-i18n="reports"',
    'data-i18n="budgets"',
    'data-i18n="ai_assistant"',
    'data-i18n="kakebo"',
]

for check in data_i18n_checks:
    test(f"  {check}", check in html)

# Test 5: Verify t() function exists and uses currentLang
print("\n5. t() function implementation:")
test("  t() function exists", "function t(key)" in html)
test("  currentLang default is 'es'", "localStorage.getItem('ff-lang') || 'es'" in html)
test("  bilingualTranslations checked first", "bilingualTranslations[key]" in html)
test("  translations.es fallback", "translations.es?.[key]" in html)

# Test 6: Verify applyTranslations is called on init
print("\n6. Translation initialization:")
test("  applyTranslations() in init()", "applyTranslations();" in html)

# Test 7: Verify no hardcoded Spanish in dynamic content
print("\n7. No hardcoded Spanish in dynamic content:")
import re

# Check for textContent = 'Spanish'
textcontent_matches = re.findall(r"textContent\s*=\s*'([^']*[áéíóúñ][^']*)'", html)
test("  No hardcoded textContent with Spanish", len(textcontent_matches) == 0)

# Check for innerHTML = 'Spanish' (excluding translation definitions)
innerhtml_matches = re.findall(r"innerHTML\s*=\s*'([^']*[áéíóúñ][^']*)'", html)
test("  No hardcoded innerHTML with Spanish", len(innerhtml_matches) == 0)

# Test 8: Verify bilingualTranslations has both ES and EN
print("\n8. Bilingual translations have both languages:")
bilingual_keys = re.findall(r"(\w+):\s*\{\s*es:\s*'([^']*)',\s*en:\s*'([^']*)'", html)
test(f"  Found {len(bilingual_keys)} bilingual translations", len(bilingual_keys) > 50)

# Check a few specific ones
for key, es_val, en_val in bilingual_keys[:5]:
    test(f"  {key}: es='{es_val}', en='{en_val}'", bool(es_val) and bool(en_val))

# Summary
print("\n" + "=" * 60)
print(f"Results: {tests_passed} passed, {tests_failed} failed")
print("=" * 60)

sys.exit(0 if tests_failed == 0 else 1)

File: test_translations_e2e.js
#!/usr/bin/env node
/**
 * E2E Translation Test for FamilyFinance
 * Simulates the actual t() function and verifies all translations work
 */
const fs = require('fs');
const path = require('path');

const htmlFile = path.join(__dirname, 'static', 'index.html');
const html = fs.readFileSync(htmlFile, 'utf8');

const scriptMatch = html.match(/<script>([\s\S]*?)<\/script>/);
if (!scriptMatch) {
  console.log('❌ No <script> found');
  process.exit(1);
}

const script = scriptMatch[1];

function extractTranslations() {
  const translations = { es: {}, en: {} };
  const bilingualTranslations = {};

  const bibStart = script.indexOf('const bilingualTranslations = {');
  const transStart = script.indexOf('const translations = {');

  if (bibStart > -1) {
    const bibEnd = script.indexOf('};', bibStart + 30);
    const bibBlock = script.substring(bibStart + 35, bibEnd);
    const bibLines = bibBlock.split('\n');
    let currentKey = null;
    let esVal = '';
    let enVal = '';

    bibLines.forEach(line => {
      const keyMatch = line.match(/^\s+([a-z_][a-z0-9_]*):\s*\{\s*es:\s*'([^']*)',\s*en:\s*'([^']*)'\s*\}/);
      if (keyMatch) {
        bilingualTranslations[keyMatch[1]] = { es: keyMatch[2], en: keyMatch[3] };
      }
    });
  }

  if (transStart > -1) {
    const esBlockMatch = script.substring(transStart).match(/es:\s*\{([\s\S]*?)\n\s*\},\s*\n\s*en:\s*\{/);
    const enBlockMatch = script.substring(transStart).match(/en:\s*\{([\s\S]*?)\n\s*\},\s*\n\s*es:\s*\{/);

    if (esBlockMatch) {
      const esLines = esBlockMatch[1].split('\n');
      esLines.forEach(line => {
        const m = line.match(/^\s+([a-z_][a-z0-9_]*):\s*'([^']*)'/);
        if (m) translations.es[m[1]] = m[2];
      });
    }

    if (enBlockMatch) {
      const enLines = enBlockMatch[1].split('\n');
      enLines.forEach(line => {
        const m = line.match(/^\s+([a-z_][a-z0-9_]*):\s*'([^']*)'/);
        if (m) translations.en[m[1]] = m[2];
      });
    }
  }

  return { translations, bilingualTranslations };
}

const { translations, bilingualTranslations } = extractTranslations();

function t(key, lang = 'es') {
  const b = bilingualTranslations[key];
  if (b && b.es) return b[lang] || b.es || key;
  return translations[lang]?.[key] || translations.es?.[key] || key;
}

let passed = 0;
let failed = 0;

function test(name, condition) {
  if (condition) {
    console.log(`  ✅ ${name}`);
    passed++;
  } else {
    console.log(`  ❌ ${name}`);
    failed++;
  }
}

console.log('='.repeat(60));
console.log('FamilyFinance Translation E2E Tests');
console.log('='.repeat(60));

console.log('\n1. Spanish translations (default language):');
const esChecks = [
  ['new_card', 'Nueva Tarjeta'],
  ['new_service', 'Nuevo Servicio'],
  ['add_income', 'Agregar Ingreso'],
  ['add_expense', 'Agregar Gasto'],
  ['expenses', 'Gastos'],
  ['income', 'Ingresos'],
  ['debt', 'Deuda'],
  ['save', 'Guardar'],
  ['loading', 'Cargando...'],
  ['password', 'Contraseña'],
  ['enter', 'Ingresar'],
  ['summary', 'Resumen'],
  ['reports', 'Reportes'],
  ['budgets', 'Presupuestos'],
  ['ai_assistant', 'Asistente IA'],
  ['kakebo', 'Kakebo'],
  ['pay_immediately', 'Pagar de Inmediato'],
  ['pay_overdue', 'Pagar Vencido'],
  ['pay_early', 'Pagar Adelantado'],
  ['no_income', 'No hay ingresos registrados'],
  ['no_expenses', 'No hay gastos registrados'],
  ['no_debts', 'No hay deudas registradas'],
  ['no_cards', 'No hay tarjetas registradas'],
  ['no_services', 'No hay servicios registrados'],
  ['upcoming_payments', 'Próximos Pagos'],
  ['kakebo_this_month', 'Kakebo - Este Mes'],
  ['needs', 'Necesidades'],
  ['wants', 'Deseos'],
  ['culture', 'Cultura'],
  ['unexpected', 'Inesperado'],
];

esChecks.forEach(([key, expected]) => {
  const actual = t(key, 'es');
  test(`${key} = "${actual}"`, actual === expected);
});

console.log('\n2. English translations:');
const enChecks = [
  ['new_card', 'New Card'],
  ['new_service', 'New Service'],
  ['add_income', 'Add Income'],
  ['add_expense', 'Add Expense'],
  ['expenses', 'Expenses'],
  ['income', 'Income'],
  ['debt', 'Debt'],
  ['save', 'Save'],
  ['loading', 'Loading...'],
  ['password', 'Password'],
  ['enter', 'Enter'],
  ['summary', 'Summary'],
  ['reports', 'Reports'],
  ['budgets', 'Budgets'],
  ['ai_assistant', 'AI Assistant'],
  ['kakebo', 'Kakebo'],
  ['pay_immediately', 'Pay Immediately'],
  ['pay_overdue', 'Pay Overdue'],
  ['pay_early', 'Pay Early'],
  ['no_income', 'No income recorded'],
  ['no_expenses', 'No expenses recorded'],
  ['no_debts', 'No debts recorded'],
  ['no_cards', 'No cards recorded'],
  ['no_services', 'No services recorded'],
  ['upcoming_payments', 'Upcoming Payments'],
  ['kakebo_this_month', 'Kakebo - This Month'],
  ['needs', 'Needs'],
  ['wants', 'Wants'],
  ['culture', 'Culture'],
  ['unexpected', 'Unexpected'],
];

enChecks.forEach(([key, expected]) => {
  const actual = t(key, 'en');
  test(`${key} = "${actual}"`, actual === expected);
});

console.log('\n3. Bilingual translations (both languages):');
const bibChecks = [
  'card_limit_usage',
  'current_balance_label',
  'view_projection_short',
  'amount_label',
  'description_label',
  'category_label',
  'date_label',
  'name_label',
  'initial_amount_label',
  'interest_rate_pct',
  'monthly_payment_label',
  'start_date_label',
  'next_payment_label',
  'limit_label',
  'actual_balance_label',
  'cutoff_day_label',
  'provider_label',
  'monthly_amount_label',
  'payment_day_label',
  'reminder_days_label',
  'save_label',
  'salary_label',
  'freelance_label',
  'investment_label',
  'gift_label',
  'other_income_label',
  'kakebo_category_label',
  'needs_label',
  'wants_label',
  'culture_label',
  'unexpected_label',
  'this_month',
  'view_all_count',
  'no_income_month',
  'no_description',
  'all_income_label',
  'view_current_month',
  'no_expenses_month',
  'all_expenses_label',
  'pay_immediately',
  'pay_overdue',
  'pay_early',
];

bibChecks.forEach(key => {
  const b = bilingualTranslations[key];
  if (b) {
    test(`${key}: es="${b.es}", en="${b.en}"`, !!b.es && !!b.en);
  } else {
    test(`${key}: exists in bilingualTranslations`, false);
  }
});

console.log('\n4. Default language is Spanish:');
const defaultLangMatch = script.match(/localStorage\.getItem\('ff-lang'\)\s*\|\|\s*'es'/);
test("currentLang default is 'es'", !!defaultLangMatch);

console.log('\n5. applyTranslations() is called on init:');
const initMatch = script.includes('async function init()') && script.includes('applyTranslations()');
test('applyTranslations() called in init()', initMatch);

console.log('\n6. data-i18n attributes in HTML:');
const dataI18nCount = (html.match(/data-i18n="/g) || []).length;
test(`${dataI18nCount} data-i18n attributes found`, dataI18nCount > 30);

console.log('\n7. No hardcoded Spanish in dynamic JS code:');
const codeLines = script.split('\n');
let hardcodedIssues = 0;
codeLines.forEach((line, i) => {
  if (line.includes('const translations') || line.includes('const bilingualTranslations')) return;
  if (line.match(/textContent\s*=\s*'[^']*[áéíóúñ][^']*'/)) hardcodedIssues++;
  if (line.match(/innerHTML\s*=\s*'[^']*[áéíóúñ][^']*'/)) hardcodedIssues++;
});
test('No hardcoded Spanish in dynamic code', hardcodedIssues === 0);

console.log('\n8. Key ES/EN translation counts:');
const esCount = Object.keys(translations.es).length;
const enCount = Object.keys(translations.en).length;
const bibCount = Object.keys(bilingualTranslations).length;
test(`translations.es has ${esCount} keys`, esCount > 100);
test(`translations.en has ${enCount} keys`, enCount > 100);
test(`bilingualTranslations has ${bibCount} keys`, bibCount > 50);

console.log('\n' + '='.repeat(60));
console.log(`Results: ${passed} passed, ${failed} failed`);
console.log('='.repeat(60));

process.exit(failed === 0 ? 0 : 1);
